Utilities/mkpack.py

```diff
@@ -0,0 +1,307 @@
+     (and, indirectly, from the Pebble apk or from libpebble2), or through
+     a "fake flash" tool that constructs a $PLATFORM_spi.bin for QEMU.
+
+  2) A C header file, which is included from RebbleOS.  Each resource
+     compiled by mkpack is given an enum entry, and the C header file
+     allows the resources to be references symbolically.
+
+  2.5) A dependency file, to be used by a build system.  Ideally, make
+       knows to ask mkpack to generate a dependency file; that way, a
+       pbpack is rebuilt every time either the input JSON changes, or any
+       of the resource pack inputs change.
+
+mkpack is smart enough to know to get resources in a handful of ways.  It
+can (aspirationally):
+
+  * Extract a raw resource from another resource pack (handy for
+    extracting fonts and such, before we have replacements of our own).
+  * Import a raw resource from disk (including a PNG, or a font file).
+  * Import an image from disk, converting and crushing to a palettized
+    PNG.
+  * Convert a graphic to system framebuffer format, for use as a splash
+    screen.
+"""
+
+__author__ = "Joshua Wise <joshua@joshuawise.com>"
+
+from stm32_crc import crc32
+import struct
+import json
+
+TAB_OFS = 0x0C
+RES_OFS = 0x200C
+
+def load_resource_from_pbpack(fname, resid):
+    """
+    Returns resource number |resid| from the pbpack file specified by
+    |fname|.
+    """
+    
+    with open(fname, 'r') as fd:
+        nrsrc = struct.unpack('I', fd.read(4))[0]
+        
+        for i in range(nrsrc):
+            fd.seek(TAB_OFS + i * 16)
+            (idx, ofs, sz, crc) = struct.unpack('iiiI', fd.read(16))
+            
+            if idx == resid:
+                break
+        
+        if idx != resid:
+            raise IOError("res %d not found in pbpack \"%s\"" % (resid, fname, ))
+        
+        fd.seek(RES_OFS + ofs)
+        return fd.read(sz)
+
+def load_resource_from_disk(fname):
+    """
+    Simply loads a file.
+    """
+    
+    with open(fname, 'r') as fd:
+        return fd.read()
+
+def save_pbpack(fname, rsrcs):
+    """
+    Outputs a handful of resources to a file.
+    
+    |rsrcs| is a list of resources, with the first mapping to resource index
+    "1".  Although the PebbleOS resource structure permits a sparse mapping
+    -- i.e., one in which one must read the whole resource table to find the
+    index that one desires -- the RebbleOS resource loader simply ignores
+    the ID number in each table entry, and indexes directly in to find what
+    it wants.  (And, further, every Pebble pbpack that I can find only has
+    them in order.)  So we, in keeping, will only generate things like that.
+    
+    """
+    
+    # First, turn the resource table into a list of entries, including
+    # index, offset, size, and CRC.
+    def mk_ent(data):
+        ent = { "idx": mk_ent.idx, "offset": mk_ent.offset, "size": len(data), "crc": crc32(data), "data": data }
```

Comment:
Remove the spaces inside the `{ ` and ` }`.

Reply:
done